### PR TITLE
Use translatable diagnostics in `rustc_const_eval`

### PR DESCRIPTION
**Pull request by fee1-dead**
This PR:

* adds a `no_span` parameter to `note` / `help` attributes when using `Subdiagnostic` to allow adding notes/helps without using a span
* has minor tweaks and changes to error messages

### CODE REVIEW

File: compiler/rustc_middle/src/mir/interpret/error.rs [hunk 1]
```diff
@@ -241,17 +274,31 @@ pub struct ScalarSizeMismatch {
...
+    AllocRange,
+}
+
 /// Error information for when the program caused Undefined Behavior.
-pub enum UndefinedBehaviorInfo {
+#[derive(Debug)]
+pub enum UndefinedBehaviorInfo<'a> {
     /// Free-form case. Only for errors that are never caught!
-    Ub(String),
+    // Ub(String),
```
Comment (RalfJung) [inline compiler/rustc_middle/src/mir/interpret/error.rs:297]: Miri still needs this, I don't think we can get away with removing it. (Same for the other free-form errors here.)

Comment (RalfJung) [reply compiler/rustc_middle/src/mir/interpret/error.rs:297]: Oh, this moved to `Custom`. Not sure how I feel about grouping free-form errors and validity failures, those are quite different and I'd rather keep them separate.

File: compiler/rustc_middle/src/mir/interpret/error.rs [hunk 2]
```diff
@@ -319,110 +363,81 @@ pub enum UndefinedBehaviorInfo {
...
+            InvalidChar(val) => I::InvalidChar { span, val: format!("{val:08x}") },
+            InvalidTag(val) => I::InvalidTag { span, val: format!("{val:x}") },
+            InvalidStr(err) => I::InvalidStr { span, err: err.to_string() },
+            InvalidUninitBytes(Some((alloc, info))) => {
+                I::InvalidUninitBytes { span, alloc, access: info.access, uninit: info.uninit }
             }
-            ValidationFailure { path: Some(path), msg } => {
-                write!(f, "constructing invalid value at {path}: {msg}")
+            ScalarSizeMismatch(self::ScalarSizeMismatch { target_size, data_size }) => {
+                I::ScalarSizeMismatch { span, target_size, data_size }
```
Comment (RalfJung) [inline compiler/rustc_middle/src/mir/interpret/error.rs:421]: There's a *ton* of redundancy and duplication in here now, as shown by these big `match`es which do nothing but translate from one enum to another. Is there any way to avoid/reduce that?

File: compiler/rustc_middle/src/mir/interpret/mod.rs [hunk 3]
```diff
@@ -34,7 +34,22 @@ macro_rules! err_ub {
 
 #[macro_export]
 macro_rules! err_ub_format {
-    ($($tt:tt)*) => { err_ub!(Ub(format!($($tt)*))) };
+    ($msg:expr $(, $($name:ident = $value:expr),* $(,)?)?) => {{
```
Comment (RalfJung) [inline compiler/rustc_middle/src/mir/interpret/mod.rs:37]: What are those `=` productions about? Looks like the macro now needs a doc comment to explain how to use it.

**Comment by RalfJung** [comment]
We have plans to change the error rendering for const-eval and Miri errors quite a bit, see https://github.com/rust-lang/miri/issues/2200. I have no idea how that interacts with these translation changes, does it make them harder or easier? I expect harder because now there is another huge new subsystem one has to learn to even being doing such a change.

File: compiler/rustc_middle/src/mir/interpret/error.rs (hunk 1 above)
Comment (fee1-dead) [reply compiler/rustc_middle/src/mir/interpret/error.rs:297]: Yeah. I should probably keep this since not having this would require porting `miri` to translatable diagnostics. Which I think should still be nice to do, but just not in this PR now.

File: compiler/rustc_const_eval/src/interpret/cast.rs [hunk 4]
```diff
@@ -138,12 +140,16 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
...
+                    let dest_bytes = dest.layout.size.bytes();
+                    let src_ty = format!("{}", src.layout.ty);
+                    let dest_ty = format!("{}", dest.layout.ty);
                     throw_ub_format!(
-                        "transmuting from {}-byte type to {}-byte type: `{}` -> `{}`",
-                        src.layout.size.bytes(),
-                        dest.layout.size.bytes(),
-                        src.layout.ty,
-                        dest.layout.ty,
+                        fluent::const_eval_invalid_transmute,
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/cast.rs:148]: If this has a dedicated message now, why should it still use the `Custom` variant?

Or conversely, if `Custom` variants are now 'typed' like this, then why do we need the other variants at all?

Also, for Miri we can still use arbitrary strings here, right?

File: compiler/rustc_const_eval/src/interpret/validity.rs [hunk 5]
```diff
@@ -397,28 +425,49 @@ impl<'rt, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> ValidityVisitor<'rt, 'mir, '
             self.path,
             AlignmentCheckFailed { required, has } =>
                 {
-                    "an unaligned {kind} (required {} byte alignment but found {})",
-                    required.bytes(),
-                    has.bytes(),
+                    found = "unaligned_ptr",
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/validity.rs:428]: How does this work, where does `unaligned_ptr` turn into something pretty?

More comments/docs seem needed here.

File: compiler/rustc_const_eval/src/interpret/validity.rs [hunk 6]
```diff
@@ -496,22 +553,28 @@ impl<'rt, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> ValidityVisitor<'rt, 'mir, '
...
                 try_validation!(
                     value.to_bool(),
                     self.path,
-                    InvalidBool(..) =>
-                        { "{:x}", value } expected { "a boolean" },
+                    InvalidBool(..) => expected("bool"),
+                    {
+                        found = "custom",
+                        value = format!("{value:x}"),
+                    }
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/validity.rs:564]: I think we need to work on the syntax here.  The order of "$value \expected a boolean" was deliberate since that is how it is printed. Also now it looks like the case ends after the `,` and then the curly brace part is completely unrelated, that's very confusing.

What do all these strings and names even mean?

Comment (fee1-dead) [reply compiler/rustc_const_eval/src/interpret/validity.rs:564]: They are located in `messages.ftl` but I agree that this should be improved. I will look into making this typed instead of using closures and macros.

File: compiler/rustc_middle/messages.ftl [hunk 7]
```diff
@@ -43,3 +43,189 @@ middle_requires_lang_item = requires `{$name}` lang_item
...
+    an error has already been reported elsewhere (this should not usually be printed)
+
+middle_size_of_unsized =
+    size_of called on unsized type `{$ty}`
+
+middle_adjust_for_foreign_abi_error =
+    target architecture {$arch} does not support `extern {$abi}` ABI
+
+middle_uninit_unsized_local =
+    unsized local is used while uninitialized
```
Comment (RalfJung) [inline compiler/rustc_middle/messages.ftl:231]: Is there a way to split the interpreter messages into their own file? Having everything in rustc_middle collected into a single file makes things rather messy; these things have little to do with each other apart from the fact that due to how the query system works they need to be in rustc_middle.

(Ideally all the interpreter stuff would be in rustc_const_eval but at least with the old error structure that wasn't possible. Maybe it is possible now, with the translations introducing an extra indirection?)

File: compiler/rustc_middle/src/mir/interpret/error.rs (hunk 2 above)
Comment (fee1-dead) [reply compiler/rustc_middle/src/mir/interpret/error.rs:421]: This is kind of hard, since the span of the diagnostic is not known until we decide to emit it. (many uses use the stack trace to find out where to point at) And this would have to either be a manual impl doing all the `span_note` and `set_arg` or using a `derive` to generate that which requires translating enums.

File: compiler/rustc_const_eval/src/errors.rs [hunk 8]
```diff
@@ -194,3 +211,195 @@ pub(crate) struct InteriorMutabilityBorrow {
...
+}
+
+#[derive(LintDiagnostic)]
+#[diag(const_eval_align_check_failed)]
+pub struct AlignmentCheckFailed {
+    pub has: u64,
+    pub required: u64,
+    #[subdiagnostic]
+    pub frames: Vec<FrameNote>,
+}
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/errors.rs:367]: Are all these types here auto-generated or do they need to be manually synced with the .flt file (in terms of field/variable names etc)?

File: compiler/rustc_middle/src/mir/interpret/error.rs (hunk 1 above)
Comment (RalfJung) [reply compiler/rustc_middle/src/mir/interpret/error.rs:297]: Yeah certainly let's leave translating Miri for later. Also I think right now this would seriously slow down Miri development (looking at this PR and the enormous overhead introduced by translations) so I am not sure it would be a good idea.

File: compiler/rustc_const_eval/messages.ftl [hunk 9]
```diff
@@ -89,3 +101,293 @@ const_eval_interior_mutable_data_refer =
...
+            [valid_char] a valid unicode scalar value (in `0..=0x10FFFF` but not in `0xD800..=0xDFFF`)
+            [float] a floating point number
+            [int] an integer
+            [init_scalar] initialized scalar value
+            [ref] a reference
+            [box] a box
+            [vtable_ptr] a vtable pointer
+            *[other] {""}
+        }
+    }
```
[the remark this answers is not in the document]
Comment (fee1-dead) [reply compiler/rustc_const_eval/messages.ftl:183]: It could be possible to mitigate this, through using the `IntoDiagnostic` traits and encoding possible variants as enums. Then it would have comments explaining what needs to be changed when adding new variants.

File: compiler/rustc_middle/src/mir/interpret/error.rs (hunk 2 above)
Comment (RalfJung) [reply compiler/rustc_middle/src/mir/interpret/error.rs:421]: Somehow for the `Custom` variant this seems to work though? Or does that vie up on some of the compile-time checks or so?

File: compiler/rustc_const_eval/src/errors.rs (hunk 8 above)
Comment (RalfJung) [reply compiler/rustc_const_eval/src/errors.rs:367]: Uff, that's suboptimal, and so easy to get wrong. I thought "there is no variable named `foo`" is an error you see at runtime in Python bot not Rust ;) . Are there plans to improve this? Currently it looks like translations introduce quite a bit of boilerplate and make the code significantly more fragile.

File: tests/ui/consts/const-eval/heap/alloc_intrinsic_nontransient_fail.stderr [hunk 10]
```diff
@@ -1,8 +1,10 @@
-error: untyped pointers are not allowed in constant
+error: unsupported untyped pointer in constant
   --> $DIR/alloc_intrinsic_nontransient_fail.rs:6:1
    |
 LL | const FOO: *const i32 = foo();
    | ^^^^^^^^^^^^^^^^^^^^^
+   |
+   = note: pointers cast from integers or from `const_allocate` are not allowed in the final value of constants
```
Comment (RalfJung) [inline tests/ui/consts/const-eval/heap/alloc_intrinsic_nontransient_fail.stderr:7]: This text isn't right. It's not about where the pointer comes from, it's about the type: if this was an `&i32` things would be fine. Also ptrs cast from int *are* allowed, e.g. `const FOO: *const i32 = 0 as *const i32` works fine.

Comment (fee1-dead) [reply tests/ui/consts/const-eval/heap/alloc_intrinsic_nontransient_fail.stderr:7]: Yeah, I was getting a bit confused about this message. oli suggested this on discord: "Only integers cast to pointers or references are supported".

But the following also emits the error: [play](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7cfd656c3a19efada59cedf88306fc34)

```
struct S;

impl Drop for S {
    fn drop(&mut self) {}
}

const X: *const S = &S;
```

Comment (fee1-dead) [reply tests/ui/consts/const-eval/heap/alloc_intrinsic_nontransient_fail.stderr:7]: Oh right, since the reference turns into a pointer for the constant it isn't allowed.

File: compiler/rustc_const_eval/src/errors.rs (hunk 8 above)
Comment (fee1-dead) [reply compiler/rustc_const_eval/src/errors.rs:367]: Yeah it's definitely increasing the work required to add a variant to `InterpError`.

I think it would be nice if there is a DSL that generated both fluent resources and the Rust types that derived `Diagnostic` and `Subdiagnostic`, but I'm not sure if that's the plan.

cc @rust-lang/wg-diagnostics any thoughts on this?

Comment (clubby789) [reply compiler/rustc_const_eval/src/errors.rs:367]: I have a PR to validate variables used open (#111269) but it unfortunately doesn't work with subdiagnostics

File: compiler/rustc_const_eval/messages.ftl (hunk 9 above)
[the remark this answers is not in the document]
Comment (clubby789) [reply compiler/rustc_const_eval/messages.ftl:183]: You could probably either duplicate the whole message as an enum, or make the `err_kind` an enum implementing `IntoDiagnosticArg`

**Comment by RalfJung** [comment]
> > does it make them harder or easier? I expect harder because now there is another huge new subsystem one has to learn to even being doing such a change.
> 
> I think I could do it in this PR by adding an enum that prints brief "title"s for errors. This PR also makes it impossible to use `InterpError` as the "title" of an error since the current change makes it so that it can only do `span_label`s or attach messages to an existing diagnostic.

Probably better to separate functional changes from major refactors like this.

However Miri currently relies quite a bit on the ability to put InterpError into the title...

File: tests/ui/consts/const-eval/heap/alloc_intrinsic_nontransient_fail.stderr (hunk 10 above)
Comment (RalfJung) [reply tests/ui/consts/const-eval/heap/alloc_intrinsic_nontransient_fail.stderr:7]: What happens is that we recursively walk the constant in a type-directed way, following the references. Once we are done, we check if there are any pointers left from the constant to some memory that has not been accounted for -- and if that is the case, we bail, since we don't have reliable type information for that memory so we don't know how to intern it.

File: compiler/rustc_middle/src/mir/interpret/error.rs (hunk 2 above)
Comment (RalfJung) [reply compiler/rustc_middle/src/mir/interpret/error.rs:421]: I think what I don't get is why this can't directly refer to `fluent::middle_scalar_size_mismatch` or something like that. There's an indirection going on with that `I::ScalarSizeMismatch` type. Is that related to the `span_note`/`set_arg` things you mentioned?

File: compiler/rustc_const_eval/src/errors.rs (hunk 8 above)
Comment (RalfJung) [reply compiler/rustc_const_eval/src/errors.rs:367]: fluent is already doing some form of code generation (right? I saw some `use` statements that look like this), so I was hoping a fluent line like
```
middle_uninit_unsized_local =
    unsized local is used while uninitialized {$variable}
```
could auto-generate a corresponding Rust struct with a field for the `variable`. Things like `fluent::const_eval_invalid_align_details` are already a thing so it seems we can refer to strings in a checked way?

File: compiler/rustc_middle/src/mir/interpret/error.rs (hunk 2 above)
Comment (fee1-dead) [reply compiler/rustc_middle/src/mir/interpret/error.rs:421]: Yeah, with translatable diagnostics these messages are supposed to exist as `struct`/`enum`s that derive the `IntoDiagnostic` trait. Manual implementations are still supposed to use `IntoDiagnostic` and those can use something like `err.span_note(span, fluent::middle_scalar_size_mismatch)`. One of my early approaches was to make `InterpError` return a `DiagnosticMessage` such that it could exist both as a label and as an error. But after that I looked at uses of InterpError in rustc and thought that its only use was as a span_label so I chose to use span_label.

That said, we can use some kind of a hack by inspecting the diagnostic after we insert the interperror label in miri so I'm not sure if we should revert back to message only

Comment (RalfJung) [reply compiler/rustc_middle/src/mir/interpret/error.rs:421]: In my view, long-term there should be two things defined for each `InterpError`: a 'title' and a 'label', with the intention that the title is rendered as the main error text and the label is rendered at the span.

I am not sure what is the best way to represent this. I had planned to just make this two separate functions that return a string each, but that was pre-translation obviously.

Comment (fee1-dead) [reply compiler/rustc_middle/src/mir/interpret/error.rs:413]: It can be done, by making both the title and the label `DiagnosticMessage`s (two functions) and have two functions that add the required arguments (non-translatable parts such as type paths) for them respectively. That could be the followup work.

File: tests/ui/consts/const-eval/heap/alloc_intrinsic_nontransient_fail.stderr (hunk 10 above)
Comment (fee1-dead) [reply tests/ui/consts/const-eval/heap/alloc_intrinsic_nontransient_fail.stderr]: Changed to "only integers cast to pointers or references are supported."

File: compiler/rustc_middle/messages.ftl (hunk 7 above)
Comment (fee1-dead) [reply compiler/rustc_middle/messages.ftl:231]: I don't think it is possible without using trait objects. The procedural macro compiles the fluent resource and generate consts that are `DiagnosticMessage`s that are then used in the `derive` macro for `IntoDiagnostic`. Therefore the crate can only refer to messages that exist in their own crate

File: compiler/rustc_middle/src/mir/interpret/error.rs (hunk 2 above)
Comment (fee1-dead) [reply compiler/rustc_middle/src/mir/interpret/error.rs:413]: I [created a thread](https://rust-lang.zulipchat.com/#narrow/stream/336883-i18n/topic/.60derive.28Subdiagnostic.29.60.2C.20but.20span.20comes.20from.20somewhere.20else) about this on Zulip. I think the easiest way forward would be to not use these traits but use inherent methods that either return the message or add arguments for formatting.

File: tests/ui/consts/const-eval/heap/alloc_intrinsic_nontransient_fail.stderr (hunk 10 above)
Comment (RalfJung) [reply tests/ui/consts/const-eval/heap/alloc_intrinsic_nontransient_fail.stderr]: I would phrase it as "memory only reachable via raw pointers is not supported"

File: compiler/rustc_middle/src/mir/interpret/error.rs (hunk 2 above)
Comment (RalfJung) [reply compiler/rustc_middle/src/mir/interpret/error.rs:413]: Yeah it is fine to do this in follow-up changes. I just mentioned it in case it would affect your design decisions now.

File: compiler/rustc_const_eval/src/interpret/validity.rs (hunk 6 above)
Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/validity.rs:564]: Also once you figured out the interface to the fluent messages we need to work on the macro syntax. So let's keep this subthread open for that purpose.

File: compiler/rustc_middle/messages.ftl (hunk 7 above)
Comment (RalfJung) [reply compiler/rustc_middle/messages.ftl:231]: Is it possible to move the entire function with the big `match` on `InterpError` that converts it to fluent data structures could be moved into rustc_const_eval? The type needs to be in rustc_middle but I believe this function could be moved. This requires an extension trait (since we can't use an inherent impl cross-crate) so we didn't consider this worth it in the past, but with all this fluent boilerplate I think the scale tips towards having this in rustc_const_eval.

Comment (fee1-dead) [reply compiler/rustc_middle/messages.ftl]: Moved.

File: compiler/rustc_const_eval/src/interpret/validity.rs (hunk 6 above)
Comment (fee1-dead) [reply compiler/rustc_const_eval/src/interpret/validity.rs]: I've replaced the previous macro stuff with enum variants which should make it very straight forward.

File: compiler/rustc_const_eval/src/interpret/validity.rs (hunk 5 above)
Comment (fee1-dead) [reply compiler/rustc_const_eval/src/interpret/validity.rs]: No more stringly typed messages. Variants are now split to make it easy to translate and read.

File: compiler/rustc_const_eval/src/interpret/intrinsics.rs [hunk 11]
```diff
@@ -513,7 +522,9 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 let op = self.eval_operand(op, None)?;
                 let cond = self.read_scalar(&op)?.to_bool()?;
                 if !cond {
-                    throw_ub_format!("`assume` called with `false`");
+                    throw_ub!(Custom(CustomSubdiagnostic::label(|| {
+                        fluent::const_eval_assume_false
+                    })));
```
Comment (oli-obk) [inline compiler/rustc_const_eval/src/interpret/intrinsics.rs:527]: how is `throw_ub_custom!` different from this?

**Review by oli-obk — CHANGES_REQUESTED** [review]
r=me after nits fixed and commits squashed

File: compiler/rustc_const_eval/src/interpret/intrinsics.rs (hunk 11 above)
Comment (fee1-dead) [reply compiler/rustc_const_eval/src/interpret/intrinsics.rs:527]: They are the same. I wrote these before `throw_ub_custom!`.

**Comment by fee1-dead** [comment]
@bors r=oli-obk

rebased

File: compiler/rustc_const_eval/src/const_eval/error.rs [hunk 12]
```diff
@@ -34,151 +70,545 @@ impl<'tcx> Into<InterpErrorInfo<'tcx>> for ConstEvalErrKind {
...
-                        for _ in 0..times {
-                            err.span_note(span, line.clone());
-                        }
-                    } else {
-                        err.span_note(
-                            span,
-                            format!("[... {} additional calls {} ...]", times, &line),
-                        );
-                    }
+impl<'a> ReportErrorExt for UndefinedBehaviorInfo<'a> {
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/const_eval/error.rs:152]: This shouldn't be in const_eval. It is also relevant for Miri.

compiler/rustc_const_eval/src/errors.rs might be a good place?

File: compiler/rustc_middle/src/mir/interpret/error.rs [hunk 13]
```diff
@@ -242,9 +244,26 @@ pub struct ScalarSizeMismatch {
...
+            }
+        }
+    )*}
+}
+
+impl_into_diagnostic_arg_through_debug! {
+    AllocId,
+    Pointer,
+    AllocRange,
+}
```
Comment (RalfJung) [inline compiler/rustc_middle/src/mir/interpret/error.rs:261]: This seems like a strange hack? Debug output should not be used for user-visible error messages...

File: compiler/rustc_mir_transform/src/const_prop_lint.rs [hunk 14]
```diff
@@ -267,14 +267,6 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
...
             Err(error) => {
                 trace!("InterpCx operation failed: {:?}", error);
-                // Some errors shouldn't come up because creating them causes
-                // an allocation, which we should avoid. When that happens,
-                // dedicated error variants should be introduced instead.
-                assert!(
-                    !error.kind().formatted_string(),
-                    "const-prop encountered formatting error: {}",
-                    error
-                );
```
Comment (RalfJung) [inline compiler/rustc_mir_transform/src/const_prop_lint.rs:277]: Why is this being removed?

File: src/tools/miri/src/diagnostics.rs [hunk 15]
```diff
@@ -302,8 +318,30 @@ pub fn report_error<'tcx, 'mir>(
 
     let stacktrace = ecx.generate_stacktrace();
     let (stacktrace, was_pruned) = prune_stacktrace(stacktrace, &ecx.machine);
-    e.print_backtrace();
-    msg.insert(0, e.to_string());
+    let (e, backtrace) = e.into_parts();
+    backtrace.print_backtrace();
+
+    let extra = match e {
```
Comment (RalfJung) [inline src/tools/miri/src/diagnostics.rs:324]: This block of code needs a comment.

(Why does this even need changing?)

File: compiler/rustc_mir_transform/src/const_prop.rs [hunk 16]
```diff
@@ -367,7 +378,7 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
                 op
             }
             Err(e) => {
-                trace!("get_const failed: {}", e);
+                trace!("get_const failed: {e:?}");
```
Comment (RalfJung) [inline compiler/rustc_mir_transform/src/const_prop.rs:381]: More suspicious debug printing. Does this regress how readable traces are?

File: compiler/rustc_middle/src/mir/interpret/mod.rs [hunk 17]
```diff
@@ -74,6 +78,11 @@ macro_rules! throw_ub {
     ($($tt:tt)*) => { do yeet err_ub!($($tt)*) };
 }
 
+#[macro_export]
+macro_rules! throw_ub_diag {
+    ($($tt:tt)*) => { do yeet err_ub_format!($($tt)*) };
```
Comment (RalfJung) [inline compiler/rustc_middle/src/mir/interpret/mod.rs:83]: I am confused by the naming here. How does this differ from `throw_ub_format`?

File: compiler/rustc_middle/src/mir/interpret/error.rs [hunk 18]
```diff
@@ -474,51 +454,26 @@ pub enum ResourceExhaustionInfo {
...
-
 /// A trait for machine-specific errors (or other "machine stop" conditions).
-pub trait MachineStopType: AsAny + fmt::Display + Send {}
+pub trait MachineStopType: Any + fmt::Debug + Send {
+    fn diagnostic_message(&self) -> DiagnosticMessage;
+    fn add_args(
+        self: Box<Self>,
+        adder: &mut dyn FnMut(Cow<'static, str>, DiagnosticArgValue<'static>),
+    );
+}
```
[the remark this answers is not in the document]
Comment (fee1-dead) [reply compiler/rustc_middle/src/mir/interpret/error.rs:465]: Done.

File: tests/ui/consts/const-eval/heap/alloc_intrinsic_nontransient_fail.stderr (hunk 10 above)
Comment (fee1-dead) [reply tests/ui/consts/const-eval/heap/alloc_intrinsic_nontransient_fail.stderr]: changed.

File: compiler/rustc_const_eval/src/interpret/cast.rs (hunk 4 above)
Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/cast.rs:148]: I'm fine with leaving FIXMEs, I was just trying to understand why the code looks the way it does. If anything I feel it'd make sense to move more messages to `Custom` than the other way around, but I might be missing some parts of the trade-off here.

File: compiler/rustc_mir_transform/src/const_prop.rs (hunk 16 above)
Comment (RalfJung) [reply compiler/rustc_mir_transform/src/const_prop.rs:381]: As someone who used these messages for debugging in the past I am certain this is worse than before -- the exact variant doesn't matter; the information about what went wrong matters. (We have RUST_CTFE_BACKTRACE=immediate to figure out where the error comes from, which is a lo more helpful than the variant anyway.) Maybe the ugly debug output is good enough but I object to attempts to selling this as an improvement.

In Miri there is some code which turns errors into a string. Is it possible to use that for a `Display` impl for `InterpError`?

Comment (fee1-dead) [reply compiler/rustc_mir_transform/src/const_prop.rs:381]: Ah I see. I could use that to print the diagnostic message if that is preferred.

Comment (RalfJung) [reply compiler/rustc_mir_transform/src/const_prop.rs]: Yeah I think that would be better than debug-printing. The code can be shared with Miri, I assume.

However this can happen in a follow-up PR. Let's get this one landed before it bitrots.

**Comment by fee1-dead** [comment]
fixed ci error

@bors r=oli-obk,RalfJung

**Comment by fee1-dead** [comment]
rebased

@bors r=oli-obk,RalfJung
